python/ray/data/datasource/parquet_datasource.py

```diff
@@ -254,6 +280,53 @@ def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
 
         return read_tasks
 
+    def _estimate_files_encoding_ratio(self) -> float:
+        """Return an estimate of the Parquet files encoding ratio.
+
+        To avoid OOMs, it is safer to return an over-estimate than an underestimate.
+        """
+        if not DatasetContext.get_current().decoding_size_estimation:
+            return PARQUET_ENCODING_RATIO_ESTIMATE_DEFAULT
+
+        # Sample a few rows from Parquet files to estimate the encoding ratio.
+        # Launch tasks to sample multiple files remotely in parallel.
+        # Evenly distributed to sample N rows in i-th row group in i-th file.
+        # TODO(ekl/cheng) take into account column pruning.
+        start_time = time.perf_counter()
+        num_files = len(self._pq_ds.pieces)
+        num_samples = int(num_files * PARQUET_ENCODING_RATIO_ESTIMATE_SAMPLING_RATIO)
+        min_num_samples = min(
+            PARQUET_ENCODING_RATIO_ESTIMATE_MIN_NUM_SAMPLES, num_files
+        )
+        max_num_samples = min(
+            PARQUET_ENCODING_RATIO_ESTIMATE_MAX_NUM_SAMPLES, num_files
+        )
+        num_samples = max(min(num_samples, max_num_samples), min_num_samples)
+
+        # Evenly distributed to choose which file to sample, to avoid biased prediction
+        # if data is skewed.
+        file_samples = [
+            self._pq_ds.pieces[idx]
+            for idx in np.linspace(0, num_files - 1, num_samples).astype(int).tolist()
```

Comment:
Question: I see that this is attempting to avoid data skew inherent in the file order (e.g. small files or files with more entropy within a given column happened to be at the front of the list), but would it be better to avoid data skew bias by ensuring that we're choosing files uniformly from the file size distribution, e.g. if we sorted the files by size before doing this `linspace` draw?

I'm not sure how good of a proxy file size skew is for data entropy skew in practice... If so, it would be interesting to look at this for a future PR.

Reply:
Yeah I am not sure how good of file size being a proxy for data skew neither. In the past, I have seen files sizes are similar, but one file's columns can be much wider than another file's. I agree we can further optimize it in the future PR if we see this pattern.